README.md
<img width="500" src="https://raw.githubusercontent.com/tfeldmann/simplematch/main/docs/simplematch.svg" alt="logo">

# simplematch

> Minimal, super readable string pattern matching for python.

[![PyPI Version][pypi-image]][pypi-url]
![PyPI - License](https://img.shields.io/pypi/l/simplematch)
[![tests](https://github.com/tfeldmann/simplematch/actions/workflows/tests.yml/badge.svg?branch=main)](https://github.com/tfeldmann/simplematch/actions/workflows/tests.yml)

```python
import simplematch

simplematch.match("He* {planet}!", "Hello World!")
>>> {"planet": "World"}

simplematch.match("It* {temp:float}°C *", "It's -10.2°C outside!")
>>> {"temp": -10.2}
```

## Installation

`pip install simplematch`

(Or just drop the `simplematch.py` file in your project.)

## Syntax

`simplematch` has only two syntax elements:

- wildcard `*`
- capture group `{name}`

Capture groups can be named (`{name}`), unnamed (`{}`) and typed (`{name:float}`).

The following types are available:

- `int`
- `float`
- `email`
- `url`
- `ipv4`
- `ipv6`
- `bitcoin`
- `ssn` (social security number)
- `ccard` (matches Visa, MasterCard, American Express, Diners Club, Discover, JCB)

For now, only named capture groups can be typed.

Then use one of these functions:

```python
import simplematch

simplematch.match(pattern, string) # -> returns a `dict` on match, `None` otherwise.
simplematch.test(pattern, string)  # -> returns `True` on match, `False` otherwise.
```

Or use a `Matcher` object:

```python
import simplematch as sm

matcher = sm.Matcher(pattern)

matcher.match(string) # -> returns a dict or None
matcher.test(string)  # -> returns True / False
matcher.regex         # -> shows the generated regex
```

## Basic usage

```python
import simplematch as sm

# extracting data
sm.match(
    pattern="Invoice_*_{year}_{month}_{day}.pdf",
    string="Invoice_RE2321_2021_01_15.pdf")
>>> {"year": "2021", "month": "01", "day": "15"}

# test match only
sm.test("ABC-{value:int}", "ABC-13")
>>> True
```

## Typed matches

```python
import simplematch as sm

matcher = sm.Matcher("{year:int}-{month:int}: {value:float}")

# extracting data
matcher.match("2021-01: -12.786")
>>> {"year": 2021, "month": 1, "value": -12.786}

# month is no integer -> no match and return `None`.
matcher.match("2021-AB: Hello")
>>> None

# no extraction, only test for match
matcher.test("1234-01: 123.123")
>>> True

# show generated regular expression
matcher.regex
>>> '^(?P<year>[+-]?[0-9]+)\\-(?P<month>[+-]?[0-9]+):\\ (?P<value>[+-]?(?:[0-9]*[.])?[0-9]+)$'

# show registered converters
matcher.converters
>>> {'year': <class 'int'>, 'month': <class 'int'>, 'value': <class 'float'>}
```

## Register your own types

You can register your own types to be available for the `{name:type}` matching syntax
with the `register_type` function.

`simplematch.register_type(name, regex, converter=str)`

- `name` is the name to use in the matching syntax
- `regex` is a regular expression to match your type
- `converter` is a callable to convert a match (`str` by default)

### Example

Register a `smiley` type to detect smileys (`:)`, `:(`, `:/`) and getting their moods:

```python
import simplematch as sm

def mood_convert(smiley):
    moods = {
        ":)": "good",
        ":(": "bad",
        ":/": "sceptic",
    }
    return moods.get(smiley, "unknown")

sm.register_type("smiley", r":[\)\(\/]", mood_convert)

sm.match("I'm feeling {mood:smiley} *", "I'm feeling :) today!")
>>> {"mood": "good"}
```

## CLI Command

You can also install `simplematch` for use as a CLI command e.g. using `pipx`.

```sh
pipx install simplematch
```

### Usage

```sh
usage: simplematch [-h] [--regex] pattern [strings ...]

positional arguments:
  pattern     A matching pattern
  strings     The string to match

options:
  -h, --help  show this help message and exit
  --regex     Show the generated regular expression
```

### Example

Extract a date from a specific file name:

```sh
simplematch "Invoice_*_{year}_{month}_{day}.pdf" "Invoice_RE2321_2021_01_15.pdf"
>>> {"year": "2021", "month": "01", "day": "15"}
```

## Background

`simplematch` aims to fill a gap between parsing with `str.split()` and regular
expressions. It should be as simple as possible, fast and stable.

The `simplematch` syntax is transpiled to regular expressions under the hood, so
matching performance should be just as good.

I hope you get some good use out of this!

## Contributions

Contributions are welcome! Just submit a PR and maybe get in touch with me via email
before big changes.

## License

[MIT](https://choosealicense.com/licenses/mit/)

<!-- Badges -->

[pypi-image]: https://img.shields.io/pypi/v/simplematch
[pypi-url]: https://pypi.org/project/simplematch/


CHANGELOG.md
# Changelog

## v1.4 (2023-10-05)

- `simplematch` now comes with a cli tool.

## v1.3 (2021-04-19)

- Adds a `case_sensitive` option to `test()` and `match()` function. (`True` by default)
- Allows editing and manual specification of `Matcher.regex` and `Matcher.converters`.

## v1.2 (2021-04-19)

- `match()` now returns an empty dict on matches with capture groups, `None` otherwise.
  In lower versions, `match()` always returned a dict.

## v1.1 (2021-04-16)

- The syntax for unnamed capture groups is now `{}`.
- add `letters` type

## v1.0 (2021-04-16)

- inital release


test_easymatch.py
import easymatch as em


def test_readme_example():
    result = em.match(
        pattern="Invoice*_{year}_{month}_{day}.pdf",
        string="Invoice_RE2321_2021_01_15.pdf",
    )
    assert result == {"year": "2021", "month": "01", "day": "15"}

    result = em.match("{year:int}-{month:int}: {value:float}", "2021-01: -12.786")
    assert result == {"year": 2021, "month": 1, "value": -12.786}


def test_readme_example_typehints():
    matcher = em.Matcher("{year:int}-{month:int}: {value:float}")

    assert matcher.match("2021-01: -12.786") == {
        "year": 2021,
        "month": 1,
        "value": -12.786,
    }
    assert matcher.match("2021-01-abc: Hello") == {}
    assert matcher.test("1234-01: 123.123") == True


def test_extended_readme_example():
    result = em.match(
        pattern="{path}/Invoice*_{year}_{month}_{day}.pdf",
        string="~/Documents/Invoices/Invoice_RE2321_2021_01_15.pdf",
    )
    assert result == {
        "path": "~/Documents/Invoices",
        "year": "2021",
        "month": "01",
        "day": "15",
    }


def test_regex():
    em.test("{file}.js", "archive.zip")
    em.test("{file}.js", "index.js")
    em.test("{folder}/{filename}.js", "foo/bar.js")
    em.test("*.{extension}", "/root/folder/file.exe")
    em.test("{folder}/{filename}.{extension}", "test/123.pdf")
    em.test("{*}/{filename}?{*}", "www.site.com/home/hello.js?p=1")


def test_simple():
    assert em.test("*.py", "hello.py")
    assert not em.test("*.zip", "hello.py")

    assert em.test("{file}.py", "hello.py")
    assert not em.test("{file}.zip", "hello.py")


def test_simple_matching():
    # should return the right match dict
    assert em.match("{folder}/{filename}.py", "home/hello.py") == {
        "folder": "home",
        "filename": "hello",
    }

    # should return empty object if no match
    assert em.match("{folder}/{filename}?{params}", "hello.js?p=1") == dict()

    # should match strings with . (dot) and ? (question mart) sights
    assert em.match("{folder}/{filename}?{params}", "home/hello.js?p=1") == dict(
        folder="home", filename="hello.js", params="p=1"
    )

    # should match wild cards
    assert em.match("*/{filename}", "home/hello.js") == dict(filename="hello.js")

    # NOT: should tolerate *{param} syntax - it acts as */{param}
    # this is different from the easypattern library!
    assert not em.match("*{filename}", "home/hello.js") == dict(filename="hello.js")

    # should save wild cards
    assert em.match("{*}/{filename}?{*}", "www.site.com/home/hello.js?p=1") == {
        0: "www.site.com/home",
        1: "p=1",
        "filename": "hello.js",
    }


def test_type_int():
    m = em.Matcher("{num:int}")
    assert m.match("123") == {"num": 123}
    assert m.match("-123") == {"num": -123}
    assert not m.match("-123.0")
    assert m.match("+123") == {"num": 123}
    assert m.match("+000123") == {"num": 123}
    assert m.match("0123") == {"num": 123}


def test_type_float():
    m = em.Matcher("{num:float}")
    assert m.match("123.4") == {"num": 123.4}
    assert m.match("-123.4") == {"num": -123.4}
    assert m.match("-123.0")
    assert m.match("+123.4") == {"num": 123.4}
    assert m.match("+000123.4") == {"num": 123.4}


def test_type_ipv4():
    m = em.Matcher("{ip:ipv4}")
    assert m.match("127.0.0.1") == {"ip": "127.0.0.1"}
    assert m.match("192.168.1.1") == {"ip": "192.168.1.1"}
    assert m.match("127.0.0.1") == {"ip": "127.0.0.1"}
    assert m.match("0.0.0.0") == {"ip": "0.0.0.0"}
    assert m.match("255.255.255.255") == {"ip": "255.255.255.255"}
    assert not m.match("256.256.256.256")
    assert not m.match("999.999.999.999")
    assert not m.match("1.2.3")
    assert m.match("1.2.3.4") == {"ip": "1.2.3.4"}


def test_type_ccard():
    m = em.Matcher("{card:ccard}")
    assert m.match("4569403961014710") == {"card": "4569403961014710"}
    assert m.match("5191914942157165") == {"card": "5191914942157165"}
    assert m.match("370341378581367") == {"card": "370341378581367"}
    assert m.match("38520000023237") == {"card": "38520000023237"}
    assert m.match("6011000000000000") == {"card": "6011000000000000"}
    assert m.match("3566002020360505") == {"card": "3566002020360505"}
    assert not m.match("1234566660000222")


def test_type_bitcoin():
    m = em.Matcher("{coin:bitcoin}")
    assert m.match("1KFHE7w8BhaENAswwryaoccDb6qcT6DbYY") == {
        "coin": "1KFHE7w8BhaENAswwryaoccDb6qcT6DbYY"
    }
    assert not m.match("loremipsum")
    assert m.match("16ftSEQ4ctQFDtVZiUBusQUjRrGhM3JYwe") == {
        "coin": "16ftSEQ4ctQFDtVZiUBusQUjRrGhM3JYwe"
    }
    assert m.match("1EBHA1ckUWzNKN7BMfDwGTx6GKEbADUozX") == {
        "coin": "1EBHA1ckUWzNKN7BMfDwGTx6GKEbADUozX"
    }
    assert not m.match("0xde0b295669a9fd93d5f28d9ec85e40f4cb697bae")
    assert m.match("bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq") == {
        "coin": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq"
    }


def test_type_url():
    m = em.Matcher("{url:url}")
    assert not m.match("abcdef")
    assert not m.match("www.whatever.com")
    assert m.match("https://github.com/geongeorge/i-hate-regex") == {
        "url": "https://github.com/geongeorge/i-hate-regex"
    }
    assert m.match("https://www.facebook.com/") == {"url": "https://www.facebook.com/"}
    assert m.match("https://www.google.com/") == {"url": "https://www.google.com/"}
    assert m.match("https://xkcd.com/2293/") == {"url": "https://xkcd.com/2293/"}
    assert not m.match("https://this-shouldn't.match@example.com")
    assert m.match("http://www.example.com/") == {"url": "http://www.example.com/"}


test_simplematch.py
import pytest  # type: ignore

import simplematch as sm


def test_readme_example_opener():
    assert sm.match("He* {planet}!", "Hello World!") == {"planet": "World"}
    assert sm.match("It* {temp:float}°C *", "It's -10.2°C outside!") == {"temp": -10.2}


def test_readme_example_basic_usage():
    result = sm.match(
        pattern="Invoice*_{year}_{month}_{day}.pdf",
        string="Invoice_RE2321_2021_01_15.pdf",
    )
    assert result == {"year": "2021", "month": "01", "day": "15"}
    assert sm.test("ABC-{value:int}", "ABC-13")


def test_readme_example_typehints():
    matcher = sm.Matcher("{year:int}-{month:int}: {value:float}")
    assert matcher.match("2021-01: -12.786") == {
        "year": 2021,
        "month": 1,
        "value": -12.786,
    }
    assert matcher.match("2021-01-abc: Hello") is None
    assert matcher.test("1234-01: 123.123") is True
    assert matcher.regex == (
        "^(?P<year>[+-]?[0-9]+)\\-(?P<month>[+-]?[0-9]+):\\ (?P<value>[+-]?"
        "(?:[0-9]*[.])?[0-9]+)$"
    )
    assert matcher.converters == {"year": int, "month": int, "value": float}


@pytest.mark.parametrize(
    "fmt, inp, result",
    (
        ("*.py", "hello.py", True),
        ("*.zip", "hello.py", False),
        ("{file}.py", "hello.py", True),
        ("{file}.zip", "hello.py", False),
        ("{folder}/{filename}.js", "foo/bar.js", True),
        ("*.{extension}", "/root/folder/file.exe", True),
        ("{folder}/{filename}.{extension}", "test/123.pdf", True),
        ("{}/{filename}?{}", "www.site.com/home/hello.js?p=1", True),
    ),
)
def test_simple(fmt, inp, result):
    assert sm.test(fmt, inp) == result


@pytest.mark.parametrize(
    "fmt, inp, test_result, match_result",
    (
        ("*.py", "hello.py", True, {}),
        ("{}.py", "hello.py", True, {0: "hello"}),
        ("*.py", "hello.__", False, None),
        ("{}.py", "hello.__", False, None),
    ),
)
def test_result(fmt, inp, test_result, match_result):
    assert sm.test(fmt, inp) == test_result
    assert sm.match(fmt, inp) == match_result


def test_unnamed_wildcards():
    assert sm.match("{} sees {}", "Tim sees Jacob") == {0: "Tim", 1: "Jacob"}


def test_simple_matching():
    # should return the right match dict
    assert sm.match("{folder}/{filename}.py", "home/hello.py") == {
        "folder": "home",
        "filename": "hello",
    }

    # should return None object if no match
    assert sm.match("{folder}/{filename}?{params}", "hello.js?p=1") is None

    # should match strings with . (dot) and ? (question mart) sights
    assert sm.match("{folder}/{filename}?{params}", "home/hello.js?p=1") == dict(
        folder="home", filename="hello.js", params="p=1"
    )

    # should match wild cards
    assert sm.match("*/{filename}", "home/hello.js") == dict(filename="hello.js")

    # NOT: should tolerate *{param} syntax - it acts as */{param}
    # this is different from the easypattern library!
    assert not sm.match("*{filename}", "home/hello.js") == dict(filename="hello.js")

    # should save wild cards
    assert sm.match("{}/{filename}?{}", "www.site.com/home/hello.js?p=1") == {
        0: "www.site.com/home",
        1: "p=1",
        "filename": "hello.js",
    }


@pytest.mark.parametrize(
    "inp, result",
    (
        ("123", {"num": 123}),
        ("-123", {"num": -123}),
        ("+123", {"num": 123}),
        ("+000123", {"num": 123}),
        ("0123", {"num": 123}),
        ("-123.0", None),
    ),
)
def test_type_int(inp, result):
    m = sm.Matcher("{num:int}")
    assert m.match(inp) == result


@pytest.mark.parametrize(
    "inp, result",
    (
        ("123.4", {"num": 123.4}),
        ("-123.4", {"num": -123.4}),
        ("+123.4", {"num": 123.4}),
        ("+000123.4", {"num": 123.4}),
        ("-123.0", {"num": -123.0}),
    ),
)
def test_type_float(inp, result):
    m = sm.Matcher("{num:float}")
    assert m.match(inp) == result


@pytest.mark.parametrize(
    "inp, result",
    (
        ("abcf123", {"chars": "abcf"}),
        ("abcf123#", {"chars": "abcf"}),
        ("ACBAAC_123", {"chars": "ACBAAC"}),
    ),
)
def test_type_letter(inp, result):
    m = sm.Matcher("{chars:letters}*")
    assert m.match(inp) == result


@pytest.mark.parametrize(
    "inp, is_bitcoin",
    (
        ("1KFHE7w8BhaENAswwryaoccDb6qcT6DbYY", True),
        ("loremipsum", False),
        ("16ftSEQ4ctQFDtVZiUBusQUjRrGhM3JYwe", True),
        ("1EBHA1ckUWzNKN7BMfDwGTx6GKEbADUozX", True),
        ("0xde0b295669a9fd93d5f28d9ec85e40f4cb697bae", False),
        ("bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", True),
    ),
)
def test_type_bitcoin(inp, is_bitcoin):
    m = sm.Matcher("{coin:bitcoin}")
    result = m.match(inp)
    if is_bitcoin:
        assert result == {"coin": inp}
    else:
        assert result is None


@pytest.mark.parametrize(
    "string",
    (
        "john@doe.com",
        "dotted.name@dotted.domain.org",
        "ug.ly-name_1@ug-ly.domain0.co.uk",
    ),
)
def test_type_email(string):
    matcher = sm.Matcher("{email:email}")
    assert matcher.test(string) is True
    assert matcher.match(string) == {"email": string}


@pytest.mark.parametrize(
    "inp, is_ssn",
    (
        ("123-45-6789", True),
        ("123 45 6789", False),
        ("333-22-4444", True),
        ("aaa-bbb-cccc", False),
        ("900-58-4564", False),
        ("999-58-4564", False),
        ("000-45-5454", False),
    ),
)
def test_type_ssn(inp, is_ssn):
    m = sm.Matcher("{n:ssn}")
    assert m.test(inp) == is_ssn


@pytest.mark.parametrize(
    "inp, result",
    (
        ("127.0.0.1", {"ip": "127.0.0.1"}),
        ("192.168.1.1", {"ip": "192.168.1.1"}),
        ("127.0.0.1", {"ip": "127.0.0.1"}),
        ("0.0.0.0", {"ip": "0.0.0.0"}),
        ("255.255.255.255", {"ip": "255.255.255.255"}),
        ("256.256.256.256", None),
        ("999.999.999.999", None),
        ("1.2.3", None),
        ("1.2.3.4", {"ip": "1.2.3.4"}),
    ),
)
def test_type_ipv4(inp, result):
    m = sm.Matcher("{ip:ipv4}")
    assert m.match(inp) == result


@pytest.mark.parametrize(
    "inp, result",
    (
        ("4569403961014710", {"card": "4569403961014710"}),
        ("5191914942157165", {"card": "5191914942157165"}),
        ("370341378581367", {"card": "370341378581367"}),
        ("38520000023237", {"card": "38520000023237"}),
        ("6011000000000000", {"card": "6011000000000000"}),
        ("3566002020360505", {"card": "3566002020360505"}),
        ("1234566660000222", None),
    ),
)
def test_type_ccard(inp, result):
    m = sm.Matcher("{card:ccard}")
    assert m.match(inp) == result


@pytest.mark.parametrize(
    "inp, is_url",
    (
        ("abcdef", False),
        ("www.whatever.com", False),
        ("https://github.com/geongeorge/i-hate-regex", True),
        ("https://www.facebook.com/", True),
        ("https://www.google.com/", True),
        ("https://xkcd.com/2293/", True),
        ("https://this-shouldn't.match@example.com", False),
        ("http://www.example.com/", True),
    ),
)
def test_type_url(inp, is_url):
    m = sm.Matcher("{url:url}")
    result = m.match(inp)
    if is_url:
        assert result == {"url": inp}
    else:
        assert result is None


def test_register_type():
    def mood_detect(smiley):
        moods = {
            ":)": "good",
            ":(": "bad",
            ":/": "sceptic",
        }
        return moods.get(smiley, "unknown")

    sm.register_type("smiley", r":[\(\)\/]", mood_detect)

    assert sm.match("I'm feeling {mood:smiley} *", "I'm feeling :) today!") == {
        "mood": "good"
    }


def test_case_sensitive():
    assert sm.test("Hello {}", "Hello World")
    assert not sm.test("hello {}", "Hello World")

    assert sm.test("Hello {}", "Hello World", case_sensitive=False)
    assert sm.test("hello {}", "Hello World", case_sensitive=False)

    # keep capture group names
    assert sm.match("HELLO {PlAnEt}", "Hello Earth", case_sensitive=False) == {
        "PlAnEt": "Earth"
    }


def test_manual_specification():
    m = sm.Matcher()
    m.regex = r"^(?P<test>\w+) \d+$"
    m.converters = {"test": str.upper}
    assert m.match("hello 123") == {"test": "HELLO"}


simplematch/converters.py
import decimal
from ipaddress import IPv4Address


class Str:
    regex = r".*"

    @staticmethod
    def to_python(value: str) -> str:
        return value


class Int:
    regex = r"[+-]?[0-9]"

    @staticmethod
    def to_python(value: str) -> int:
        return int(value)


class Float:
    regex = r"[+-]?([0-9]*[.])?[0-9]+"

    @staticmethod
    def to_python(value: str) -> float:
        return float(value)


class Decimal(Float):
    @staticmethod
    def to_python(value: str) -> decimal.Decimal:
        return decimal.Decimal(value)


class FourDigitYear(Int):
    regex = "[0-9]{4}"

    @staticmethod
    def to_python(value: str) -> int:
        return int(value)


class Letters(Str):
    regex = r"[a-zA-Z]+"


class RomanNumeral(Int):
    regex = r"M{0,3}(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})"


class Bitcoin(Str):
    regex = r"(bc1|[13])[a-zA-HJ-NP-Z0-9]{25,39}"


class Email(Str):
    regex = r"[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+"


class Url(Str):
    regex = (
        r"https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b"
        r"([-a-zA-Z0-9()!@:%_\+.~#?&\/\/=]*)"
    )


class IpV4:
    regex = (
        r"(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)"
        r"(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}"
    )

    def to_python(self, value) -> IPv4Address:
        return IPv4Address(value)


class IpV6:
    regex = (
        r"(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA"
        r"-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){"
        r"1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3"
        r"}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0"
        r"-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:"
        r"(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5"
        r"]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0"
        r"-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,"
        r"3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))"
    )


class Port:
    regex = (
        r"((6553[0-5])|(655[0-2][0-9])|(65[0-4][0-9]{2})|(6[0-4][0-9]{3})|"
        r"([1-5][0-9]{4})|([0-5]{0,5})|([0-9]{1,4}))"
    )


class MacAddress:
    regex = r"[a-fA-F0-9]{2}(:[a-fA-F0-9]{2}){5}"


class SocialSecurityNumber(Str):
    regex = r"(?!0{3})(?!6{3})[0-8]\d{2}-(?!0{2})\d{2}-(?!0{4})\d{4}"


class CreditCard:
    regex = (
        r"(^4[0-9]{12}(?:[0-9]{3})?$)|(^(?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6]["
        r"0-9]{2}|27[01][0-9]|2720)[0-9]{12}$)|(3[47][0-9]{13})|(^3(?:0[0-5]|[68][0-9])"
        r"[0-9]{11}$)|(^6(?:011|5[0-9]{2})[0-9]{12}$)|(^(?:2131|1800|35\d{3})\d{11}$)"
    )


class LatLon:
    regex = r"((\-?|\+?)?\d+(\.\d+)?),\s*((\-?|\+?)?\d+(\.\d+)?)"


class SemanticVersion:
    regex = (
        r"(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)"
        r"(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)"
        r"(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
        r"(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?"
    )


class JiraIssueTicket:
    regex = r"[A-Z]{2,}-\d+"


class Hashtag:
    regex = r"#[^ !@#$%^&*(),.?\":{}|<>]*"


simplematch.py
#!/usr/bin/env python3
"""
simplematch
"""
import re
from collections import namedtuple

# taken from the standard re module - minus "*{}", because that's our own syntax
SPECIAL_CHARS = {i: "\\" + chr(i) for i in b"()[]?+-|^$\\.&~# \t\n\r\v\f"}

# a regex that ensures all groups to be non-capturing. Otherwise they would appear in
# the matches
TYPE_CLEANUP_REGEX = re.compile(r"(?<!\\)\((?!\?)")

# `types` is the dict of known types that is filled with register_type
Type = namedtuple("Type", "regex converter")
types = {}


def register_type(name, regex, converter=str):
    """register a type to be available for the {value:type} matching syntax"""
    cleaned = TYPE_CLEANUP_REGEX.sub("(?:", regex)
    types[name] = Type(regex=cleaned, converter=converter)


# include some useful basic types
register_type("int", r"[+-]?[0-9]+", int)
register_type("float", r"[+-]?([0-9]*[.])?[0-9]+", float)
register_type("letters", r"[a-zA-Z]+")

# found on https://ihateregex.io/
register_type("bitcoin", r"(bc1|[13])[a-zA-HJ-NP-Z0-9]{25,39}")
register_type("email", r"[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+")
register_type("ssn", r"(?!0{3})(?!6{3})[0-8]\d{2}-(?!0{2})\d{2}-(?!0{4})\d{4}")
register_type(
    "ipv4",
    (
        r"(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]"
        r"?)){3}"
    ),
)
register_type(
    "url",
    (
        r"https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA"
        r"-Z0-9()!@:%_\+.~#?&\/\/=]*)"
    ),
)

register_type(
    # Visa, MasterCard, American Express, Diners Club, Discover, JCB
    "ccard",
    (
        r"(^4[0-9]{12}(?:[0-9]{3})?$)|(^(?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6]["
        r"0-9]{2}|27[01][0-9]|2720)[0-9]{12}$)|(3[47][0-9]{13})|(^3(?:0[0-5]|[68][0-9])"
        r"[0-9]{11}$)|(^6(?:011|5[0-9]{2})[0-9]{12}$)|(^(?:2131|1800|35\d{3})\d{11}$)"
    ),
)
register_type(
    "ipv6",
    (
        r"(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA"
        r"-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){"
        r"1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3"
        r"}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0"
        r"-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:"
        r"(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5"
        r"]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0"
        r"-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,"
        r"3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))"
    ),
)


class Matcher:
    def __init__(self, pattern="*", case_sensitive=True):
        self.converters = {}
        self.pattern = pattern
        self.case_sensitive = case_sensitive
        self.regex = self._create_regex(pattern)

    def test(self, string):
        match = self._regex_compiled.match(string)
        return match is not None

    def match(self, string):
        match = self._regex_compiled.match(string)
        if match:
            # assemble result dict
            result = match.groupdict()
            for i, x in enumerate(self._grouplist(match)):
                result[i] = x

            # run converters
            for key, converter in self.converters.items():
                result[key] = converter(result[key])
            return result
        return None

    @property
    def regex(self):
        return self._regex

    @regex.setter
    def regex(self, value):
        self._regex = value
        flags = 0 if self.case_sensitive else re.IGNORECASE
        # cache the compiled regex
        self._regex_compiled = re.compile(value, flags=flags)

    def _field_repl(self, matchobj):
        # field with type annotation
        match = re.search(r"\{(\w+):(\w+)\}", matchobj.group(0))
        if match:
            name, type_ = match.groups()
            # register this field to convert it later
            self.converters[name] = types[type_].converter
            return r"(?P<%s>%s)" % (name, types[type_].regex)

        # field without type annotation
        match = re.search(r"\{(\w+)\}", matchobj.group(0))
        if match:
            name = match.group(1)
            return r"(?P<%s>.*)" % name

    def _create_regex(self, pattern):
        self.converters.clear()  # empty converters
        result = pattern.translate(SPECIAL_CHARS)  # escape special chars
        result = result.replace("*", r".*")  # handle wildcard
        result = re.sub(r"\{\}", r"(.*)", result)  # handle unnamed group
        result = re.sub(r"\{([^\}]*)\}", self._field_repl, result)  # handle named group
        return r"^%s$" % result

    @staticmethod
    def _grouplist(match):
        """extract unnamed match groups"""
        # https://stackoverflow.com/a/53385788/300783
        named = match.groupdict()
        ignored_groups = set()
        for name, index in match.re.groupindex.items():
            if name in named:  # check twice if it is really the named attribute
                ignored_groups.add(index)
        return [
            group
            for i, group in enumerate(match.groups())
            if i + 1 not in ignored_groups
        ]

    def __repr__(self):
        return '<Matcher("%s")>' % self.pattern


def test(pattern, string, case_sensitive=True):
    return Matcher(pattern, case_sensitive=case_sensitive).test(string)


def match(pattern, string, case_sensitive=True):
    return Matcher(pattern, case_sensitive=case_sensitive).match(string)


def to_regex(pattern):
    return Matcher(pattern).regex


def simplematch_cli():
    import json
    from argparse import ArgumentParser

    parser = ArgumentParser()
    parser.add_argument("pattern", help="A matching pattern")
    parser.add_argument("strings", help="The string to match", nargs="*")
    parser.add_argument(
        "--regex", action="store_true", help="Show the generated regular expression"
    )
    args = parser.parse_args()

    m = Matcher(args.pattern)
    if args.regex:
        print(f"Regex: {m.regex}")
    for string in args.strings:
        print(json.dumps(m.match(string)))


if __name__ == "__main__":
    simplematch_cli()
